steward/src/grpc_client.rs

```diff
@@ -0,0 +1,158 @@
+use std::{thread, time::Duration};
+use steward_proto::{
+    self,
+    steward::{
+        contract_call_client::ContractCallClient as ContractClient, SubmitRequest, SubmitResponse,
+    },
+};
+use tonic::transport::{Channel, ClientTlsConfig, Uri};
+
+pub type ContractCallClient = ContractClient<Channel>;
+pub type QueryClient = QClient<Channel>;
+
+pub struct GrpcClient {
+    pub grpc_address: String,
+    pub steward_endpoints: Vec<String>,
+    pub tls_configuration: ClientTlsConfig,
+}
+
+impl GrpcClient {
+    pub fn check_for_grpc_address(&self) -> Result<(), Error> {
+        if self.grpc_address.is_empty() {
+            return Err(ErrorKind::GrpcError
+                .context("no grpc address available")
+                .into());
+        }
+
+        Ok(())
+    }
+
+    pub fn check_for_steward_addresses(&self) -> Result<(), Error> {
+        if self.steward_endpoints.is_empty() {
+            return Err(ErrorKind::GrpcError
+                .context("no grpc address available")
+                .into());
+        }
+
+        Ok(())
+    }
+
+    pub async fn get_query_client(&self) -> Result<QueryClient, Error> {
+        self.check_for_grpc_address()?;
+
+        QueryClient::connect(self.grpc_address.clone())
+            .await
+            .map_err(|e| ErrorKind::GrpcError.context(e).into())
+    }
+
+    pub async fn get_contract_clients(&self) -> Result<Vec<ContractCallClient>, Error> {
+        self.check_for_steward_addresses()?;
+
+        let mut clients = Vec::new();
+
+        for addr in &self.steward_endpoints {
+            let endpoint = Channel::builder(addr.parse::<Uri>().unwrap())
+                .tls_config(self.tls_configuration.clone())?;
+
+            let client = match ContractCallClient::connect(endpoint).await {
+                Ok(res) => res,
+                Err(err) => return Err(ErrorKind::GrpcError.context(err).into()),
+            };
+
+            clients.push(client);
+        }
+
+        Ok(clients)
+    }
+
+    pub async fn submit_request(
+        &self,
+        request: SubmitRequest,
+    ) -> Result<(Vec<SubmitResponse>, Vec<Error>), Error> {
+        // First verify voting period has started
+        let mut query_client = self.get_query_client().await?;
+
+        loop {
+            let commit_period_response = match query_client
+                .query_commit_period(QueryCommitPeriodRequest {})
+                .await
+            {
+                Ok(res) => res.into_inner(),
+                Err(status) => return Err(ErrorKind::GrpcError.context(status).into()),
+            };
+
+            if commit_period_response.current_height >= commit_period_response.vote_period_start
+                && commit_period_response.current_height < commit_period_response.vote_period_end
+            {
+                break;
+            }
+
+            // Wait a few seconds before retrying
+            thread::sleep(Duration::from_secs(10));
```

Comment:
Don't you think it'll be good if we can use `tokio::time` here so that we can make a timeout for our loop? Something like this: 

https://github.com/PeggyJV/gravity-bridge/blob/16bf5569ec490fe333b99cc36231021a3b96a134/orchestrator/gorc/src/commands/deploy/erc20.rs#L103